File: app/src/main/java/org/michaelbel/moviemade/ui/modules/movie/views/RatingView.java
package org.michaelbel.moviemade.ui.modules.movie.views;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.LinearLayout;

import org.michaelbel.moviemade.R;
import org.michaelbel.moviemade.utils.Theme;
import org.michaelbel.moviemade.utils.LayoutHelper;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.ContextCompat;

public class RatingView extends LinearLayout {

    private StarView[] stars = new StarView[5];

    public RatingView(Context context) {
        super(context, null);
        initialize(context);
    }

    public RatingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs, 0);
        initialize(context);
    }

    public RatingView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize(context);
    }

    private void initialize(Context context) {
        setOrientation(HORIZONTAL);

        int starCount = 5;
        for (int i = 0; i < starCount; i++) {
            StarView view = new StarView(context);
            view.setLayoutParams(LayoutHelper.makeLinear(20, 20));
            addView(view);
            stars[i] = view;
        }
    }

    public void setRating(float rating) {
        float myRating = rating / 2;

        int a = (int) myRating;
        int b = (int) (10 * (myRating - a));
        int c;
        c = (b >= 5) ? 5 : 0;

        if (c == 0) {
            switch (a) {
                case 0:
                    for(StarView star : stars) {
                        star.setIcon(StarView.ICON_STAR_BORDER);
                    }
                    break;
                case 1:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR_BORDER);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 2:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 3:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 4:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 5:
                    for(StarView star : stars) {
                        star.setIcon(StarView.ICON_STAR);
                    }
                    break;
            }
        } else if (c == 5) {
            switch (a) {
                case 0:
                    stars[0].setIcon(StarView.ICON_STAR_HALF);
                    stars[1].setIcon(StarView.ICON_STAR_BORDER);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 1:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR_HALF);
                    stars[2].setIcon(StarView.ICON_STAR_BORDER);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 2:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR_HALF);
                    stars[3].setIcon(StarView.ICON_STAR_BORDER);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 3:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR_HALF);
                    stars[4].setIcon(StarView.ICON_STAR_BORDER);
                    break;
                case 4:
                    stars[0].setIcon(StarView.ICON_STAR);
                    stars[1].setIcon(StarView.ICON_STAR);
                    stars[2].setIcon(StarView.ICON_STAR);
                    stars[3].setIcon(StarView.ICON_STAR);
                    stars[4].setIcon(StarView.ICON_STAR_HALF);
                    break;
            }
        }
    }

    private class StarView extends AppCompatImageView {

        public static final int ICON_STAR = 0;
        public static final int ICON_STAR_HALF = 1;
        public static final int ICON_STAR_BORDER = 2;

        private Drawable icon;

        public StarView(Context context) {
            super(context);
        }

        public void setIcon(int style) {
            if (style == ICON_STAR) {
                icon = Theme.getIcon(R.drawable.ic_star, ContextCompat.getColor(getContext(), R.color.accent_yellow));
            } else if (style == ICON_STAR_HALF) {
                icon = Theme.getIcon(R.drawable.ic_star_half, ContextCompat.getColor(getContext(), R.color.accent_yellow));
            } else if (style == ICON_STAR_BORDER) {
                icon = Theme.getIcon(R.drawable.ic_star_border, ContextCompat.getColor(getContext(), R.color.accent_yellow));
            }

            setImageDrawable(icon);
        }
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/movie/views/RatingView.kt
package org.michaelbel.moviemade.presentation.features.movie.views

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.core.content.ContextCompat
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.core.LayoutHelper
import org.michaelbel.moviemade.core.ViewUtil

class RatingView: LinearLayoutCompat {

    companion object {
        const val ICON_STAR = 0
        const val ICON_STAR_HALF = 1
        const val ICON_STAR_BORDER = 2
    }

    private val stars = arrayOfNulls<StarView>(5)

    constructor(context: Context): super(context, null) {
        initialize(context)
    }

    constructor(context: Context, attrs: AttributeSet?): super(context, attrs, 0) {
        initialize(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        initialize(context)
    }

    private fun initialize(context: Context) {
        orientation = LinearLayoutCompat.HORIZONTAL

        val starCount = 5
        for (i in 0 until starCount) {
            val view = StarView(context)
            view.layoutParams = LayoutHelper.makeLinear(context,20, 20)
            addView(view)
            stars[i] = view
        }
    }

    fun setRating(rating: Float) {
        val myRating = rating / 2

        val a = myRating.toInt()
        val b = (10 * (myRating - a)).toInt()
        val c: Int
        c = if (b >= 5) 5 else 0

        if (c == 0) {
            when (a) {
                0 -> for (star in stars) {
                    star?.setIcon(ICON_STAR_BORDER)
                }
                1 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR_BORDER)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                2 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                3 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                4 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                5 -> for (star in stars) {
                    star?.setIcon(ICON_STAR)
                }
            }
        } else {
            when (a) {
                0 -> {
                    stars[0]?.setIcon(ICON_STAR_HALF)
                    stars[1]?.setIcon(ICON_STAR_BORDER)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                1 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR_HALF)
                    stars[2]?.setIcon(ICON_STAR_BORDER)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                2 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR_HALF)
                    stars[3]?.setIcon(ICON_STAR_BORDER)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                3 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR_HALF)
                    stars[4]?.setIcon(ICON_STAR_BORDER)
                }
                4 -> {
                    stars[0]?.setIcon(ICON_STAR)
                    stars[1]?.setIcon(ICON_STAR)
                    stars[2]?.setIcon(ICON_STAR)
                    stars[3]?.setIcon(ICON_STAR)
                    stars[4]?.setIcon(ICON_STAR_HALF)
                }
            }
        }
    }

    private inner class StarView(context: Context): AppCompatImageView(context) {

        private var icon: Drawable? = null

        fun setIcon(style: Int) {
            when (style) {
                ICON_STAR -> icon = ViewUtil.getIcon(context, R.drawable.ic_star, ContextCompat.getColor(context, R.color.accent_yellow))
                ICON_STAR_HALF -> icon = ViewUtil.getIcon(context, R.drawable.ic_star_half, ContextCompat.getColor(context, R.color.accent_yellow))
                ICON_STAR_BORDER -> icon = ViewUtil.getIcon(context, R.drawable.ic_star_border, ContextCompat.getColor(context, R.color.accent_yellow))
            }

            setImageDrawable(icon)
        }
    }
}